plugins/win-asio/asio-device.c

```diff
@@ -0,0 +1,1260 @@
+				  long granularity)
+{
+	if (!dev)
+		return;
+
+	if (granularity >= 0) {
+		if (granularity < 16)
+			granularity = 16;
+		// ensure that the buffer size is a multiple of 16, customary with ASIO drivers
+		int start = (min_size + 15) & ~15;
+		if (granularity > start)
+			start = granularity;
+
+		for (int i = start; i <= max_size && i <= 6400; i += granularity) {
+			long step = granularity * (i / granularity);
+			da_push_back(dev->buffer_sizes, &step);
+		}
+
+	} else {
+		for (int i = 0; i < 18; ++i) {
+			int size = 1 << i;
+			if (size >= min_size && size <= max_size)
+				da_push_back(dev->buffer_sizes, &size);
+		}
+	}
+}
+
+ASIOError asio_device_refresh_buffer_sizes(struct asio_device *dev)
+{
+	if (!dev || !dev->asio)
+		return ASE_NotPresent;
+
+	ASIOError err = dev->asio->lpVtbl->getBufferSize(dev->asio, &dev->min_buffer_size, &dev->max_buffer_size,
+							 &dev->preferred_buffer_size, &dev->buffer_granularity);
+
+	if (err == ASE_OK) {
+		da_clear(dev->buffer_sizes);
+		asio_device_add_buffer_sizes(dev, dev->min_buffer_size, dev->max_buffer_size,
+					     dev->preferred_buffer_size, dev->buffer_granularity);
+	}
+
+	return err;
+}
+
+int asio_device_get_preferred_buffer_size(struct asio_device *dev)
+{
+	if (!dev || !dev->asio)
+		return 0;
+	int min_buffer_size = 0;
+	int max_buffer_size = 0;
+	int buffer_granularity = 0;
+	long preferred = 0;
+	if (dev->asio->lpVtbl->getBufferSize(dev->asio, &min_buffer_size, &max_buffer_size, &preferred,
+					     &buffer_granularity) == ASE_OK) {
+		return preferred;
+	}
+	return 0;
+}
+
+int asio_device_read_buffer_size(struct asio_device *dev, int requested_size)
+{
+	if (!dev || !dev->asio)
+		return requested_size;
+
+	dev->min_buffer_size = 0;
+	dev->max_buffer_size = 0;
+	dev->buffer_granularity = 0;
+	long new_preferred = 0;
+
+	if (dev->asio->lpVtbl->getBufferSize(dev->asio, &dev->min_buffer_size, &dev->max_buffer_size, &new_preferred,
+					     &dev->buffer_granularity) == ASE_OK) {
+		if (dev->preferred_buffer_size != 0 && new_preferred != 0 &&
+		    new_preferred != dev->preferred_buffer_size)
+			dev->should_use_preferred_size = true;
+
+		if (requested_size < dev->min_buffer_size || requested_size > dev->max_buffer_size)
+			dev->should_use_preferred_size = true;
+
+		dev->preferred_buffer_size = new_preferred;
+	}
+
+	// Workaround for buggy drivers which crash if you make dynamic changes to the buffer size
+	if (strstr(dev->device_name, "Digidesign"))
+		dev->should_use_preferred_size = true;
+
+	if (dev->should_use_preferred_size) {
+		info("Using preferred size for buffer..");
+		ASIOError err = asio_device_refresh_buffer_sizes(dev);
+
+		if (err == ASE_OK) {
+			requested_size = dev->preferred_buffer_size;
+		} else {
+			requested_size = 1024;
+			warn("getBufferSize1 failed, using 1024 as fallback");
+		}
+
+		dev->should_use_preferred_size = false;
+	}
+
+	return requested_size;
+}
+
+//=============================================================
```

Comment:
If the file is large/complex enough to require comments like this to "segment off" parts of it, then the file should be split up into separate source files instead.